# IBM Data Science Professional Certificate
# Applied Data Science Capstone
## Week 3
## Segmenting and Clustering Neighborhoods in Toronto

## Part 1 - Parsing Toronto poastal codes from Wikipedia
Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M

* The dataframe will consist of three columns: `PostalCode`, `Borough`, and `Neighborhood`
* Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
* More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11  in the above table.
* If a cell has a borough but a Not assigned  neighborhood, then the neighborhood will be the same as the borough.
* Clean your Notebook and add Markdown cells to explain your work and any assumptions you are making.
* In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.

In [41]:
import numpy as np
import pandas as pd

import requests
import urllib.request
from urllib.request import urlopen

from bs4 import BeautifulSoup

url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
html = urlopen(url) 
soup = BeautifulSoup(html, 'html.parser')

Now we will scrape the table we need from the source page. There are multiple tables on the webpage, we need the one with `wikitable sortable` class.

In [42]:
soup = soup.find('table', attrs={'class':'wikitable sortable'})
table_str = str(soup.extract())
df = pd.read_html(table_str)[0]
df.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

As per bullet point to we need to ignore *Not assigned* rows. Also, according to the exercise requirements *Postal Code* must be named without space.

In [43]:
# drop the rows with Not assigned values
df_dropna = df[df.Borough != 'Not assigned'].reset_index(drop=True)

# rename the Postal Code column
df_dropna.rename(columns={'Postal Code' : 'PostalCode'}, inplace=True)
df = df_dropna
df

PostalCode           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
..         ...               ...   
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [44]:
df.shape

(103, 3)

## Part 2 - Get the latitude and the longitude coordinate

Now that we have built a dataframe of the postal code of each neighborhood along with the borough name and neighborhood name, in order to utilize the Foursquare location data, we need to get the latitude and the longitude coordinates of each neighborhood. 

As instructed, we can either use the CSV file with the coordinates of the Geocoder Python package for this task: https://geocoder.readthedocs.io/index.html.

The sequence of steps are:
* get the coordinates (either way) and save to a dataframe
* merge the dataframe with the coordinates with the postal codes

In [45]:
# read the coordinates from the csv file that has been provided
geospatial_df = pd.read_csv("https://cocl.us/Geospatial_data")
geospatial_df.rename(columns={'Postal Code': 'PostalCode'}, inplace=True)
geospatial_df.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

In [46]:
# merge the dataframes
df = pd.merge(df, geospatial_df, on='PostalCode')
df

PostalCode           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
..         ...               ...   
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

## Part 3 - Explore and cluster the neighborhoods in Toronto

You can decide to work with only boroughs that contain the word Toronto and then replicate the same analysis we did to the New York City data. It is up to you. 
* to add enough Markdown cells to explain what you decided to do and to report any observations you make. 
* to generate maps to visualize your neighborhoods and how they cluster together. 

Sequence of analysis steps:
* we will perform some steps to explore the dataset
* we will select the boroughs that have *Toronto* in their names
* define Foursquare credentials 
* start exploring the neighborhoods

First, let's get the geographical coordinates of the city of Toronto

In [125]:
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values 

import folium # map rendering library

from pandas.io.json import json_normalize

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

# import k-means from clustering stage
from sklearn.cluster import KMeans

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

In [54]:
# Let's get the geographical coordinates of Manhattan.
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [61]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df['Latitude'], df['Longitude'], df['Borough'], df['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Now we need to look at only the boroughs that have the word *Toronto* in them

In [65]:
# create dataframe with boroughs containing the word 'Toronto' in it
df_toronto = df[df.Borough.str.contains('Toronto')]
df_toronto.reset_index(inplace=True, drop=True)
df_toronto

PostalCode           Borough  \
0         M5A  Downtown Toronto   
1         M7A  Downtown Toronto   
2         M5B  Downtown Toronto   
3         M5C  Downtown Toronto   
4         M4E      East Toronto   
5         M5E  Downtown Toronto   
6         M5G  Downtown Toronto   
7         M6G  Downtown Toronto   
8         M5H  Downtown Toronto   
9         M6H      West Toronto   
10        M5J  Downtown Toronto   
11        M6J      West Toronto   
12        M4K      East Toronto   
13        M5K  Downtown Toronto   
14        M6K      West Toronto   
15        M4L      East Toronto   
16        M5L  Downtown Toronto   
17        M4M      East Toronto   
18        M4N   Central Toronto   
19        M5N   Central Toronto   
20        M4P   Central Toronto   
21        M5P   Central Toronto   
22        M6P      West Toronto   
23        M4R   Central Toronto   
24        M5R   Central Toronto   
25        M6R      West Toronto   
26        M4S   Central Toronto   
27        M5S  Downtown Toronto   
28        M6S      West Toronto   
29        M4T   Central Toronto   
30        M5T  Downtown Toronto   
31        M4V   Central Toronto   
32        M5V  Downtown Toronto   
33        M4W  Downtown Toronto   
34        M5W  Downtown Toronto   
35        M4X  Downtown Toronto   
36        M5X  Downtown Toronto   
37        M4Y  Downtown Toronto   
38        M7Y      East Toronto   

                                        Neighbourhood   Latitude  Longitude  
0                           Regent Park, Harbourfront  43.654260 -79.360636  
1         Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
2                            Garden District, Ryerson  43.657162 -79.378937  
3                                      St. James Town  43.651494 -79.375418  
4                                         The Beaches  43.676357 -79.293031  
5                                         Berczy Park  43.644771 -79.373306  
6                                  Central Bay Street  43.657952 -79.387383  
7                                            Christie  43.669542 -79.422564  
8                            Richmond, Adelaide, King  43.650571 -79.384568  
9                        Dufferin, Dovercourt Village  43.669005 -79.442259  
10  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
11                           Little Portugal, Trinity  43.647927 -79.419750  
12                       The Danforth West, Riverdale  43.679557 -79.352188  
13           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
14       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191  
15                     India Bazaar, The Beaches West  43.668999 -79.315572  
16                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
17                                    Studio District  43.659526 -79.340923  
18                                      Lawrence Park  43.728020 -79.388790  
19                                           Roselawn  43.711695 -79.416936  
20                                   Davisville North  43.712751 -79.390197  
21    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307  
22                      High Park, The Junction South  43.661608 -79.464763  
23                  North Toronto West, Lawrence Park  43.715383 -79.405678  
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
25                             Parkdale, Roncesvalles  43.648960 -79.456325  
26                                         Davisville  43.704324 -79.388790  
27                     University of Toronto, Harbord  43.662696 -79.400049  
28                                 Runnymede, Swansea  43.651571 -79.484450  
29                        Moore Park, Summerhill East  43.689574 -79.383160  
30          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
31  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049  
32  CN Tower, King and Spadi

Now let's update our map with only these 39 boroughs

In [67]:
# create a map instance
map_toronto = folium.Map(location=[location.latitude, location.longitude], zoom_start=11)

# Add markers for all 38 neighborhoods
for _, row in df_toronto.iterrows():
    label = '{} ({}), {}'.format(
        row.PostalCode, row.Neighbourhood, row.Borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [row.Latitude, row.Longitude],
        radius=5,
        popup=label,
        color='purple',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto) 
    
map_toronto

In [ ]:
#Define Foursquare Credentials and Version
CLIENT_ID = '2NOBVJ3XJE2WC0Q35EPX3JELGHNB42IX5LYOAFTVM4RWCRM0' # your Foursquare ID
CLIENT_SECRET = 'PAOPLFN0IQ42WGH21EXCPVKGCL5H0VBLWYCHTDGCFBV0K4RO' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

## Part 3 continued - Exploring the neighborhoods in Toronto

We will start with *Regent Park, Harbourfront* as it the first neighborhood in our dataframe

In [71]:
# start exploring the neighborhoods
neighborhood_name = df_toronto.loc[0, 'Neighbourhood']
neighborhood_latitude = df_toronto.loc[0, 'Latitude']
neighborhood_longitude = df_toronto.loc[0, 'Longitude']

print("Latitude and longitude of neighborhood '{}' are [{}, {}]".format(
    neighborhood_name, neighborhood_latitude, neighborhood_longitude))

Latitude and longitude of neighborhood 'Regent Park, Harbourfront' are [43.6542599, -79.3606359]


### Get the top 100 venues in 'Regent Park, Harbourfront ' within a radius of 500 meters.

In [72]:
# Construct the URL
limit = 100
radius = 500
explore_url_prefix = 'https://api.foursquare.com/v2/venues/explore'
url = '{}?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    explore_url_prefix, CLIENT_ID, CLIENT_SECRET, VERSION, 
    neighborhood_latitude, neighborhood_longitude, radius, limit)

In [ ]:
# send the GET request and examine the resutls
results = requests.get(url).json()
#results

In [75]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [ ]:
# Explore the venues
venues = results['response']['groups'][0]['items']
venues

In [84]:
#
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

<ipython-input-84-4767924a7cda>:2: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  nearby_venues = json_normalize(venues) # flatten JSON


name           categories        lat        lng
0        Roselle Desserts               Bakery  43.653447 -79.362017
1           Tandem Coffee          Coffee Shop  43.653559 -79.361809
2  Cooper Koo Family YMCA  Distribution Center  43.653249 -79.358008
3      Morning Glory Cafe       Breakfast Spot  43.653947 -79.361149
4     Body Blitz Spa East                  Spa  43.654735 -79.359874

In [86]:
# And how many venues were returned by Foursquare?
nearby_venues.shape

(46, 4)

### What type of venues are nearby?

In [88]:
print("The Venue Categories are", nearby_venues['categories'].unique())

The Venue Categories are ['Bakery' 'Coffee Shop' 'Distribution Center' 'Breakfast Spot' 'Spa'
 'Restaurant' 'Park' 'Gym / Fitness Center' 'Historic Site'
 'Chocolate Shop' 'Farmers Market' 'Dessert Shop' 'Pub'
 'Performing Arts Venue' 'French Restaurant' 'Yoga Studio' 'Café'
 'Theater' 'Event Space' 'Shoe Store' 'Italian Restaurant' 'Brewery'
 'Art Gallery' 'Cosmetics Shop' 'Electronics Store' 'Beer Store' 'Bank'
 'Hotel' 'Wine Shop' 'Antique Shop']


Let's use the function to explore all the neighborhoods

In [89]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [91]:
#Get venues for all neighborhoods in our dataset
toronto_venues = getNearbyVenues(names=df_toronto['Neighbourhood'], latitudes=df_toronto['Latitude'], longitudes=df_toronto['Longitude'])

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
R

In [92]:
toronto_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3      Morning Glory Cafe       43.653947       -79.361149   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3       Breakfast Spot  
4                  Spa

In [93]:
toronto_venues.shape

(1610, 7)

In [96]:
print('There are {} unique venues in these neighborhoods.'.format(toronto_venues['Venue'].nunique()))
print('And there are {} venue categories.'.format(toronto_venues['Venue Category'].nunique()))

There are 1039 unique venues in these neighborhoods.
And there are 235 venue categories.


In our 39 boroughs that we looked at, there are 1039 unique venues in these neighborhoods and there are 235 venue categories.

How many how many venues are there per neighborhood?

In [97]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            59   
Brockton, Parkdale Village, Exhibition Place                           23   
Business reply mail Processing Centre, South Ce...                     16   
CN Tower, King and Spadina, Railway Lands, Harb...                     17   
Central Bay Street                                                     61   
Christie                                                               16   
Church and Wellesley                                                   78   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             33   
Davisville North                                                        7   
Dufferin, Dovercourt Village                                           14   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         5   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          25   
India Bazaar, The Beaches West                                         20   
Kensington Market, Chinatown, Grange Park                              66   
Lawrence Park                                                           4   
Little Portugal, Trinity                                               43   
Moore Park, Summerhill East                                             3   
North Toronto West, Lawrence Park                                      20   
Parkdale, Roncesvalles                                                 14   
Queen's Park, Ontario Provincial Government                            33   
Regent Park, Harbourfront                                              46   
Richmond, Adelaide, King                                               96   
Rosedale                                                                4   
Roselawn                                                                3   
Runnymede, Swansea                                                     36   
St. James Town                                                         79   
St. James Town, Cabbagetown                                            41   
Stn A PO Boxes                                                         98   
Studio District                                                        37   
Summerhill West, Rathnelly, South Hill, Forest ...                     14   
The Annex, North Midtown, Yorkville                                    19   
The Beaches                                                             5   
The Danforth West, Riverdale                                           43   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         32   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             59   
Brockton, Parkdale Village, Exhibition Place                            23   
Business reply mail Processing Centre, South Ce...                      16   
CN Tower, King and Spadina, Railway Lands, Harb...                      17   
Central Bay Street                                                      61   
Christie                                                                16   
Church and Wellesley                                                    78   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              33   
Davisville North                    

### What venue categories are there?

In [99]:
print("The Venue Categories are", toronto_venues['Venue Category'].unique())

The Venue Categories are ['Bakery' 'Coffee Shop' 'Distribution Center' 'Breakfast Spot' 'Spa'
 'Restaurant' 'Park' 'Gym / Fitness Center' 'Historic Site'
 'Chocolate Shop' 'Farmers Market' 'Dessert Shop' 'Pub'
 'Performing Arts Venue' 'French Restaurant' 'Yoga Studio' 'Café'
 'Theater' 'Event Space' 'Shoe Store' 'Italian Restaurant' 'Brewery'
 'Art Gallery' 'Cosmetics Shop' 'Electronics Store' 'Beer Store' 'Bank'
 'Hotel' 'Wine Shop' 'Antique Shop' 'Portuguese Restaurant' 'Beer Bar'
 'Creperie' 'Mexican Restaurant' 'Sushi Restaurant' 'Diner' 'Hobby Shop'
 'Burrito Place' 'Fried Chicken Joint' 'Nightclub' 'Discount Store'
 'Japanese Restaurant' 'Fast Food Restaurant' 'Smoothie Shop'
 'Sandwich Place' 'Gym' 'College Auditorium' 'Bar' 'Clothing Store'
 'Pizza Place' 'Comic Shop' 'Plaza' 'Ramen Restaurant' 'Burger Joint'
 'Music Venue' 'Thai Restaurant' 'College Rec Center' 'Shopping Mall'
 'Tanning Salon' 'Steakhouse' 'Gastropub' 'Sporting Goods Shop'
 'Bookstore' 'Lake' 'Modern European 

### What is the most frequent venue category in each of the neighborhoods?

For this we will create a dataframe with the most frequent types of venues for each of the 39 boroughs.

This task involves several steps:
* encode categoric variables into numeric
* group rows by neighborhood and by taking the mean of the frequency of occurrence of each category
* put that into a pandas dataframe
* create the new dataframe and display the top 10 venues for each neighborhood.

In [101]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0            0        0                   0             0               0   
1            0        0                   0             0               0   
2            0        0                   0             0               0   
3            0        0                   0             0               0   
4            0        0                   0             0               0   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                0                 0                    0             0   
1                0                 0                    0             0   
2                0                 0                    0             0   
3                0                 0                    0             0   
4                0                 0                    0             0   

   Aquarium  ...  Theme Restaurant  Tibetan Restaurant  Toy / Game Store  \
0         0  ...                 0                   0                 0   
1         0  ...                 0                   0                 0   
2         0  ...                 0                   0                 0   
3         0  ...                 0                   0                 0   
4         0  ...                 0                   0                 0   

   Trail  Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0      0              0                              0                 0   
1      0              0                              0                 0   
2      0              0                              0                 0   
3      0              0                              0                 0   
4      0              0                              0                 0   

   Vietnamese Restaurant  Wine Bar  Wine Shop  
0                      0         0          0  
1                      0         0          0  
2                      0         0          0  
3                      0         0          0  
4                      0         0          0  

[5 rows x 235 columns]

In [102]:
toronto_onehot.shape

(1610, 235)

In [103]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio   Airport  \
0                                         Berczy Park     0.000000  0.000000   
1        Brockton, Parkdale Village, Exhibition Place     0.000000  0.000000   
2   Business reply mail Processing Centre, South C...     0.062500  0.000000   
3   CN Tower, King and Spadina, Railway Lands, Har...     0.000000  0.058824   
4                                  Central Bay Street     0.016393  0.000000   
5                                            Christie     0.000000  0.000000   
6                                Church and Wellesley     0.025641  0.000000   
7                      Commerce Court, Victoria Hotel     0.000000  0.000000   
8                                          Davisville     0.000000  0.000000   
9                                    Davisville North     0.000000  0.000000   
10                       Dufferin, Dovercourt Village     0.000000  0.000000   
11             First Canadian Place, Underground city     0.000000  0.000000   
12    Forest Hill North & West, Forest Hill Road Park     0.000000  0.000000   
13                           Garden District, Ryerson     0.000000  0.000000   
14  Harbourfront East, Union Station, Toronto Islands     0.000000  0.000000   
15                      High Park, The Junction South     0.000000  0.000000   
16                     India Bazaar, The Beaches West     0.000000  0.000000   
17          Kensington Market, Chinatown, Grange Park     0.000000  0.000000   
18                                      Lawrence Park     0.000000  0.000000   
19                           Little Portugal, Trinity     0.023256  0.000000   
20                        Moore Park, Summerhill East     0.000000  0.000000   
21                  North Toronto West, Lawrence Park     0.050000  0.000000   
22                             Parkdale, Roncesvalles     0.000000  0.000000   
23        Queen's Park, Ontario Provincial Government     0.030303  0.000000   
24                          Regent Park, Harbourfront     0.021739  0.000000   
25                           Richmond, Adelaide, King     0.000000  0.000000   
26                                           Rosedale     0.000000  0.000000   
27                                           Roselawn     0.000000  0.000000   
28                                 Runnymede, Swansea     0.027778  0.000000   
29                                     St. James Town     0.000000  0.000000   
30                        St. James Town, Cabbagetown     0.024390  0.000000   
31                                     Stn A PO Boxes     0.010204  0.000000   
32                                    Studio District     0.027027  0.000000   
33  Summerhill West, Rathnelly, South Hill, Forest...     0.000000  0.000000   
34                The Annex, North Midtown, Yorkville     0.000000  0.000000   
35                                        The Beaches     0.000000  0.000000   
36                       The Danforth West, Riverdale     0.023256  0.000000   
37           Toronto Dominion Centre, Design Exchange     0.000000  0.000000   
38                     University of Toronto, Harbord     0.031250  0.000000   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0             0.000000      0.000000        0.000000         0.000000   
1             0.000000      0.000000        0.000000         0.000000   
2             0.000000      0.000000        0.000000         0.000000   
3             0.058824      0.058824        0.117647         0.176471   
4             0.000000      0.000000        0.000000         0.000000   
5             0.000000      0.000000        0.000000         0.000000   
6             0.000000      0.000000        0.000000         0.000000   
7             0.000000      0.000000        0.000000         0.000000   
8             0.000000      0.000000        0.000000         0.000000   
9             0.000000      0.000000        0.000000         0.000000   
10            0.000000      0.000000 

In [105]:
toronto_grouped.shape

(39, 235)

### Let's print each neighborhood along with the top 5 most common venues and then put that into a pandas dataframe

In [106]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.08
1        Cocktail Bar  0.05
2      Farmers Market  0.03
3            Beer Bar  0.03
4  Seafood Restaurant  0.03


----Brockton, Parkdale Village, Exhibition Place----
                    venue  freq
0                    Café  0.13
1          Breakfast Spot  0.09
2             Coffee Shop  0.09
3            Climbing Gym  0.04
4  Furniture / Home Store  0.04


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
                venue  freq
0  Light Rail Station  0.12
1         Yoga Studio  0.06
2       Auto Workshop  0.06
3          Skate Park  0.06
4          Restaurant  0.06


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
              venue  freq
0   Airport Service  0.18
1    Airport Lounge  0.12
2  Airport Terminal  0.12
3          Boutique  0.06
4           Airport  0.06


----Central Bay Street----


In [107]:
# First, let's write a function to sort the venues in descending order.
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [138]:
# Now let's create the new dataframe and display the top 10 venues for each neighborhood.
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...    Light Rail Station   
3  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar                Bakery        Farmers Market   
1        Breakfast Spot           Coffee Shop                   Gym   
2           Yoga Studio         Garden Center               Brewery   
3        Airport Lounge      Airport Terminal           Coffee Shop   
4        Sandwich Place    Italian Restaurant                  Café   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0              Beer Bar            Restaurant    Seafood Restaurant   
1                Bakery               Stadium         Burrito Place   
2        Farmers Market  Fast Food Restaurant         Burrito Place   
3       Harbor / Marina                 Plane   Rental Car Location   
4           Salad Place       Bubble Tea Shop          Burger Joint   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0           Cheese Shop                Bistro     Basketball Stadium  
1            Restaurant          Climbing Gym              Pet Store  
2            Restaurant         Auto Workshop                 Garden  
3      Sculpture Garden              Boutique          Boat or Ferry  
4   Japanese Restaurant       Thai Restaurant     Miscellaneous Shop

## Part 3 continued - Clustering neighborhoods

Run k-means to cluster the neighborhood into 5 clusters.

In [140]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([4, 4, 4, 4, 4, 4, 4, 4, 4, 2])

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [141]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df_toronto

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighbourhood')

toronto_merged.head() # check the last columns!

PostalCode           Borough                                Neighbourhood  \
0        M5A  Downtown Toronto                    Regent Park, Harbourfront   
1        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2        M5B  Downtown Toronto                     Garden District, Ryerson   
3        M5C  Downtown Toronto                               St. James Town   
4        M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               4           Coffee Shop   
1  43.662301 -79.389494               4           Coffee Shop   
2  43.657162 -79.378937               4           Coffee Shop   
3  43.651494 -79.375418               4           Coffee Shop   
4  43.676357 -79.293031               4      Asian Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bakery                  Park                  Café   
1      Sushi Restaurant                   Gym  Fast Food Restaurant   
2        Clothing Store   Japanese Restaurant                  Café   
3                  Café             Gastropub          Cocktail Bar   
4                   Pub     Health Food Store                 Trail   

   5th Most Common Venue 6th Most Common Venue      7th Most Common Venue  \
0                    Pub        Breakfast Spot                    Theater   
1  Portuguese Restaurant                  Park                  Nightclub   
2        Bubble Tea Shop        Cosmetics Shop  Middle Eastern Restaurant   
3    American Restaurant        Lingerie Store           Department Store   
4    Distribution Center          Dessert Shop         Dim Sum Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0             Wine Shop           Event Space  Performing Arts Venue  
1    Mexican Restaurant   Japanese Restaurant     Italian Restaurant  
2                 Hotel        Lingerie Store            Pizza Place  
3              Creperie        Cosmetics Shop     Seafood Restaurant  
4                 Diner        Discount Store       Doner Restaurant

Finally, let's visualize the resulting clusters

In [142]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Examining Clusters

#### Cluster 0

In [128]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
19  Central Toronto               0        Ice Cream Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
19          Home Service                Garden      Department Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
19  Ethiopian Restaurant           Escape Room     Electronics Store   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
19  Eastern European Restaurant            Donut Shop       Doner Restaurant

#### Cluster 1

In [129]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
29   Central Toronto               1                  Park   
33  Downtown Toronto               1                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
29                 Trail          Tennis Court         Deli / Bodega   
33            Playground                 Trail          Dance Studio   

   5th Most Common Venue        6th Most Common Venue  \
29           Escape Room            Electronics Store   
33     Electronics Store  Eastern European Restaurant   

          7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
29  Eastern European Restaurant            Donut Shop      Doner Restaurant   
33                   Donut Shop      Doner Restaurant               Dog Run   

   10th Most Common Venue  
29                Dog Run  
33    Distribution Center

#### Cluster 2

In [130]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
20  Central Toronto               2     Food & Drink Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
20                  Park           Pizza Place        Breakfast Spot   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
20                 Hotel      Department Store        Sandwich Place   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
20             Wine Shop    Dim Sum Restaurant                  Diner

#### Cluster 3

In [131]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
18  Central Toronto               3                  Park   
21  Central Toronto               3                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
18    Dim Sum Restaurant           Swim School              Bus Line   
21                 Trail         Jewelry Store      Sushi Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
18           Event Space           Escape Room     Electronics Store   
21              Bus Line             Wine Shop          Dessert Shop   

          8th Most Common Venue        9th Most Common Venue  \
18  Eastern European Restaurant                   Donut Shop   
21            Electronics Store  Eastern European Restaurant   

   10th Most Common Venue  
18       Doner Restaurant  
21             Donut Shop

#### Cluster 4

In [132]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0   Downtown Toronto               4           Coffee Shop   
1   Downtown Toronto               4           Coffee Shop   
2   Downtown Toronto               4           Coffee Shop   
3   Downtown Toronto               4           Coffee Shop   
4       East Toronto               4      Asian Restaurant   
5   Downtown Toronto               4           Coffee Shop   
6   Downtown Toronto               4           Coffee Shop   
7   Downtown Toronto               4         Grocery Store   
8   Downtown Toronto               4           Coffee Shop   
9       West Toronto               4                Bakery   
10  Downtown Toronto               4           Coffee Shop   
11      West Toronto               4                   Bar   
12      East Toronto               4      Greek Restaurant   
13  Downtown Toronto               4           Coffee Shop   
14      West Toronto               4                  Café   
15      East Toronto               4  Fast Food Restaurant   
16  Downtown Toronto               4           Coffee Shop   
17      East Toronto               4           Coffee Shop   
22      West Toronto               4    Mexican Restaurant   
23   Central Toronto               4           Coffee Shop   
24   Central Toronto               4        Sandwich Place   
25      West Toronto               4        Breakfast Spot   
26   Central Toronto               4        Sandwich Place   
27  Downtown Toronto               4                  Café   
28      West Toronto               4                  Café   
30  Downtown Toronto               4                  Café   
31   Central Toronto               4           Coffee Shop   
32  Downtown Toronto               4       Airport Service   
34  Downtown Toronto               4           Coffee Shop   
35  Downtown Toronto               4           Coffee Shop   
36  Downtown Toronto               4           Coffee Shop   
37  Downtown Toronto               4           Coffee Shop   
38      East Toronto               4    Light Rail Station   

   2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
0                 Bakery                   Park                  Café   
1       Sushi Restaurant                    Gym  Fast Food Restaurant   
2         Clothing Store    Japanese Restaurant                  Café   
3                   Café              Gastropub          Cocktail Bar   
4                    Pub      Health Food Store                 Trail   
5           Cocktail Bar                 Bakery        Farmers Market   
6         Sandwich Place     Italian Restaurant                  Café   
7                   Café                   Park    Athletics & Sports   
8                   Café             Restaurant                   Gym   
9               Pharmacy            Music Venue                  Bank   
10              Aquarium                  Hotel                  Café   
11           Coffee Shop       Asian Restaurant            Restaurant   
12           Coffee Shop     Italian Restaurant        Ice Cream Shop   
13                 Hotel                   Café           Salad Place   
14        Breakfast Spot            Coffee Shop                   Gym   
15                   Gym            Pizza Place            Board Shop   
16            Restaurant                   Café                 Hotel   
17               Brewery              Gastropub                Bakery   
22                  Café        Thai Restaurant        Discount Store   
23   Sporting Goods Shop         Clothing Store  Gym / Fitness Center   
24                  Café            Coffee Shop        History Museum   
25             Gift Shop           Dessert Shop      Cuban Restaurant   
26           Pizza Place           Dessert Shop                   Gym   
27                Bakery              Bookstore                   Bar   
28           Coffee Shop                    Pub           Pizza Place   
30    

### Final Discussion

* Cluster 4 has the most number of neighborhoods while other 4 clusters have maximum 2 neighborhoods. Looks like these neighborhoods are alike due to high number of cafes and coffee shops. While cluster 0 to 3 are close to parks and trails. 
* Interestingly, Cluster 0 has Ethiopina Restaurant on Top 5 and Eastern-European Restaurant on 8th most common venues which might signal to cultural diversity in this neighborhood
* Cluster 1 combines neighborhoods good for outdoor sports activities due to its amount of parks, trails, playgrounds and dance studios
* Hotels seem to be clustered in Davisville North neighborhoods (cluster 2)

### Further Steps

* We could experiment more with the cluster size and apply techniques such as the elbow method to find the optimal number of clusters
